---
title: "GLMER Analysis Scripts"
author: "Cassandra Buzby"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#Load packages
require(ggplot2)
require(tidyr)
require(dplyr)
require(reshape2)
require(data.table)
require(doParallel)
#require(scales)
#require(stringr)
require(cybrBSA)
require(lme4)

```

```{r, eval = FALSE}
#to install cybrBSA, use:
devtools::install_github("cbuzby/cybrBSA")
```

```{r functions, include = FALSE, eval = FALSE}

cybrInputGATKTable2 <- function(rawData, yeast = TRUE){

  require(dplyr)
  require(doParallel)
  require(foreach)

  HNGLCDRXY <- read.table(rawData, header = TRUE)

  #Identify the unique values besides AD/DP/GQ/PL
  gsub(".AD", "",
       gsub(".GQ", "",
            gsub(".DP","",
                 gsub(".PL","",
                      colnames(select(HNGLCDRXY, -CHROM, -POS, -REF, -ALT)))))) %>% unique() -> Samples
  #i <- Samples[1]

  resultscdf <- foreach(i=Samples,.combine=rbind) %dopar% {
    mydf <- HNGLCDRXY %>% select(CHROM, POS, REF, ALT) %>% mutate(Dataset = i)
    AD <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("AD"))
    GQ <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("GQ"))
    DP <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("DP"))
    PL <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("PL"))
    cbind(mydf, AD , GQ , DP, PL) -> mydftotal
    colnames(mydftotal) <- c(colnames(mydf), "AD", "GQ", "DP", "PL")

    mydftotal %>% separate(AD, c('AD.REF','AD.ALT'), extra='drop') %>%
      separate(PL, c('PL.REF','PL.ALT'), extra='drop') %>%
      #Added 10/18/23:
      select(CHROM, POS,REF,ALT,Dataset,AD.REF,AD.ALT,GQ,DP,PL.REF,PL.ALT) -> mycdf

    mycdf %>% filter(grepl(",", ALT)) %>% 
      separate(ALT, c("A1", "A2"), extra = 'merge') %>%
      separate(AD.ALT, c("AD1", "AD2"), extra = 'merge') %>%
      separate(PL.ALT, c("P1", "P2"), extra = 'merge') %>%
      
      pivot_longer(c(A1, A2), names_to = "NumAlt", values_to = "ALT") %>%
      pivot_longer(c(AD1, AD2), names_to = "NumADAlt", values_to = "AD.ALT") %>%
      pivot_longer(c(P1, P2), names_to = "NumPL", values_to = "PL.ALT") %>%
      mutate(NumAlt = gsub("A", "", NumAlt),
             NumADAlt = gsub("AD", "", NumADAlt),
             NumPL = gsub("P", "", NumPL)) %>%
      filter(NumAlt == NumPL,
             NumPL == NumADAlt) %>%
      select(CHROM, POS,REF,ALT,Dataset,AD.REF,AD.ALT,GQ,DP,PL.REF,PL.ALT) -> doublecdf
    
    doublecdf %>% filter(grepl(",", ALT)) %>%
      separate(ALT, c("A1", "A2"), extra = 'merge') %>%
      separate(AD.ALT, c("AD1", "AD2"), extra = 'merge') %>%
      separate(PL.ALT, c("P1", "P2"), extra = 'merge') %>%
      
      pivot_longer(c(A1, A2), names_to = "NumAlt", values_to = "ALT") %>%
      pivot_longer(c(AD1, AD2), names_to = "NumADAlt", values_to = "AD.ALT") %>%
      pivot_longer(c(P1, P2), names_to = "NumPL", values_to = "PL.ALT") %>%
      mutate(NumAlt = gsub("A", "", NumAlt),
             NumADAlt = gsub("AD", "", NumADAlt),
             NumPL = gsub("P", "", NumPL)) %>%
      filter(NumAlt == NumPL,
             NumPL == NumADAlt) %>%
      select(CHROM, POS,REF,ALT,Dataset,AD.REF,AD.ALT,GQ,DP,PL.REF,PL.ALT) -> triplecdf
    
    rbind(mycdf, doublecdf, triplecdf) -> newcdf
    
    newcdf
  }

  if(yeast == TRUE){
    ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII",
                                         "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"),
                         CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10",
                                   "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6",
                                   "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5",
                                   "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))

    resultscdf %>% left_join(.,ChromKey) %>% select(-CHROM) %>% mutate(CHROM = chromosomes) %>% select(-chromosomes) -> results
  }else{
    results <- resultscdf
  }
  return(results)

}

cybr_weightedgauss <- function(myx){
  myy <- dnorm(1:length(myx), mean = length(myx)/2, sd = 10)
  return(weighted.mean(x = myx, y = myy, na.rm = TRUE))
} 

glmer_cb2_short <- function (..., W, formula, numgroups = FALSE, outputlength = 4, 
    return = c("Z")) 
{
    data <- list(...)
    
    require(lme4)
    if (is.null(W) || is.null(formula)) {
        stop("Weights (W) and formula must be provided")
    }
    glm_formula <- as.formula(formula)
    if (!all(names(data) %in% all.vars(glm_formula))) {
        stop("One or more variables in the formula are not provided as arguments")
    }
    for (i in all.vars(glm_formula)) {
        if (length(unique(as.data.frame(data)[, i])) < 2) {
            output <- rep(NA, outputlength)
            return(output)
        }
    }
    glm_fit <- glmer(glm_formula, data = as.data.frame(data), weights = W, 
        family = binomial)
    if (return %in% "Z") {
        output <- summary(glm_fit)$coefficients[((length(summary(glm_fit)$coefficients) * 
            0.5) + 1):((length(summary(glm_fit)$coefficients) * 
            0.75))]
    }
    if (length(output) == outputlength) {
        return(output)
    }
    else {
        return(rep(NA, outputlength))
    }
}

```

# Pre-processing

## Load in table and match to oak or wine parents

```{r, warning=FALSE, message=FALSE, eval = FALSE}

setwd("Input/")

parentSNPids <- cybrConvertParentalAlleles(ParentFiles = c("Wine_VCF.txt", "Oak_VCF.txt"), Truncate = TRUE)
parentSNPids %>% group_by(CHROM, POS) %>% summarize(Type = Type, Unique = length(unique(Type))) -> pSNPs

head(pSNPs)

```

## Exclude duplicates 

```{r, message = FALSE, warning=FALSE, eval = FALSE}

myfile <- "Input/AllCuSO4.REF_.SortedCat.vcf.output.table"

#GET RID OF EXCLUDED ONES
read.table(myfile, header = TRUE) -> alldata
alldata %>% select(CHROM, POS, ALT) %>% separate(ALT, into = c("First", "Second"), sep = ",") %>% na.omit() -> exclude

ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII",
                                         "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"),
                         CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10",
                                   "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6",
                                   "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5",
                                   "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))

exclude %>% left_join(.,ChromKey) %>% select(-CHROM) %>% mutate(CHROM = chromosomes) %>% select(-chromosomes) -> exclude

exclude %>% mutate(locus = paste(CHROM, POS)) -> exclKEY

cybrInputGATKTable2(myfile) %>% mutate(Coverage = as.numeric(AD.REF) + as.numeric(AD.ALT)) %>%
  select(POS, CHROM, Dataset, GQ, AD.REF, AD.ALT, Coverage) -> rawdata


saveRDS(rawdata, file = "Output/rawdata.rds")

################################################################################
rawdata %>% #head(3000) %>%
  merge(parentSNPids) %>% mutate(REFW = as.numeric(Type == "Wine"), REFO = as.numeric(Type == "Oak")) %>%
  group_by(Dataset, CHROM, POS) %>%
  mutate(Wine = max(REFW * as.numeric(AD.REF), REFO * as.numeric(AD.ALT)),
         Oak = max(REFW * as.numeric(AD.ALT), REFO * as.numeric(AD.REF))) %>%
  select(Dataset, POS, CHROM, Coverage, Wine, Oak) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "Reads") -> rawdata_called

rawdata_called %>% mutate(locus = paste(CHROM, POS)) %>% filter(locus %in% exclKEY$locus == FALSE) %>% select(-locus) -> rawdata_called_ex


################################################################################

saveRDS(rawdata_called_ex, file = "Output/rawdata_called_ex.rds")

head(rawdata_called_ex)
```

### Smooth Data

Data is saved under G200 for Gaussian kernel, n = 200.

```{r, eval = FALSE}
rawdata_called_ex %>% ungroup() %>% 
  group_by(Dataset, CHROM, Allele) %>%
  arrange(POS) %>%
  reframe(POS = POS, SmoothCount = frollapply(Reads, n = 200, FUN = cybr_weightedgauss, align = "center")) %>% mutate(smooth = 200) -> rawdata_called_ex_G200

saveRDS(rawdata_called_ex_G200, file = "Output/rawdata_called_ex_G200.rds")
```

```{r, echo = FALSE}
rawdata_called_ex_G200 <- readRDS("Output/rawdata_called_ex_G200.rds")
head(rawdata_called_ex_G200)
```

## Separate into bulks by experiment name

```{r, message = FALSE}

rawdata_called_ex_G200 %>% mutate(Dataset = gsub("n01_", "", Dataset)) %>%
  mutate(Dataset = gsub(".fastq", "", Dataset)) %>%
  separate(Dataset, into = c("Pool", "Dataset"), sep = "_", extra = "merge") -> rd

expnames <- read.csv("Input/experiment_names_524.csv")

rd %>% left_join(expnames) -> rd_named

```


## Set factor contrasts

```{r, message = FALSE}
#Clean data from other sequencing runs
rd_named %>% filter(Bulk != "Fluconazole") %>%
  select(-Selection, -Replicate) %>% na.omit() %>%
  mutate_if(is.character, as.factor) -> rd_factor

rd_factor %>% filter(Dataset %in% c("O1_aD", "W1_aD") == FALSE) -> rd_factor

#Set contrasts
contrasts(rd_factor$Background) <- matrix(c(0.5, -0.5))
contrasts(rd_factor$Bulk) <- matrix(c(1, 0))

saveRDS(rd_factor, "Output/rd_factor_G200.rds")

head(rd_factor)
```

################################################################################

# Run Logistic Regressions

### GLMER scripts

GLMER scripts will add in the replicate as a random effect, whereas glm will ignore replicates and average the effects across experiments. GLM will be *much* faster than GLMER. Both scripts are included, but data for the glmer() functions are saved in github, whereas data for glm is not.

```{r, message = FALSE}
#rd_factor <- readRDS("Output/rd_factor_G200.rds")

################################################################################
rd_factor %>% filter(CHROM != "M", CHROM != "I") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  filter(Pool %in% c("HNGLVDRXY", "HKTFTDRX2")) %>%
  group_by(CHROM, POS) %>% summarize(zscore = glmer_cb2_short(Bulk = Bulk,  
                                                            Background = Background,
                                                            Rep = Dataset,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Background + (1 | Rep) "),
                                                                  label = c("cept","Bulk", "Background", "Interaction"))  -> CuSO4_chr1_glmer

saveRDS(CuSO4_chr1_glmer, file = "Output/CuSO4_chr1_glmer.rds")

################################################################################
rd_factor %>% filter(CHROM != "M", CHROM != "VIII")%>% distinct() %>% 
  #Select sequencing runs for Chr VIII fixed
  filter(Pool %in% c("HVYTYDRX2","HTG3TDMXY")) %>%
  group_by(CHROM, POS) %>% summarize(zscore = glmer_cb2_short(Bulk = Bulk,  
                                                            Background = Background,
                                                            Rep = Dataset,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Background + (1 | Rep)"),
                                                                  label = c("cept","Bulk", "Background", "Interaction"))  -> CuSO4_chr8_glmer

saveRDS(CuSO4_chr8_glmer, file = "Output/CuSO4_chr8_glmer.rds ")

head(CuSO4_chr8_glmer)
```

### GLM scripts (faster, less precise)

Permutation scripts are shown for glmer and not glm.

```{r, eval = FALSE}
#rd_factor <- readRDS("Output/rd_factor_G200.rds")

################################################################################
rd_factor %>% filter(CHROM != "M", CHROM != "I") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  filter(Pool %in% c("HNGLVDRXY", "HKTFTDRX2")) %>%
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_short(Bulk = Bulk,  
                                                            Background = Background,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Background"),
                                                                  label = c("cept","Bulk", "Background", "Interaction"))  -> CuSO4_chr1_glm

saveRDS(CuSO4_chr1_glm, file = "Output/CuSO4_chr1_glm.rds")

################################################################################
rd_factor %>% filter(CHROM != "M", CHROM != "VIII")%>% distinct() %>% 
  #Select sequencing runs for Chr VIII fixed
  filter(Pool %in% c("HVYTYDRX2","HTG3TDMXY")) %>%
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_short(Bulk = Bulk,  
                                                            Background = Background,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Background"),
                                                                  label = c("cept","Bulk", "Background", "Interaction"))  -> CuSO4_chr8_glm

saveRDS(CuSO4_chr8_glm, file = "Output/CuSO4_chr8_glm.rds ")
```

## Permutations for significance

Load in data
```{r}
#CuSO4_chr1_glmer_1byRep <- readRDS("Output/CuSO4_chr1_glmer.rds")
#CuSO4_chr8_glmer_1byRep <- readRDS("Output/CuSO4_chr8_glmer.rds")
```

Permute non-selected (dilute) bulks for each pool twice
```{r}
rd_factor %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool, Dataset) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Background = Background) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Selected") -> rd_shuffled_selected

rd_factor %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool, Dataset) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Background = Background) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Dilute") %>%
  rbind(rd_shuffled_selected) -> Perm1

```

Set contrasts 
```{r}
Perm1 %>% ungroup() %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount") %>% unnest() %>%
  mutate(POS = as.numeric(POS)) %>%
  mutate_if(is.character, as.factor) -> Perm1contrasts

contrasts(Perm1contrasts$Background) <- matrix(c(-0.5, 0.5))
contrasts(Perm1contrasts$Bulk)

```

Run glmer() for each fixed chromosome
```{r}

Perm1contrasts %>%  filter(Pool %in% c("HNGLVDRXY", "HKTFTDRX2")) %>% distinct() %>% 
  
  group_by(CHROM, POS) %>% reframe(zscore = glmer_cb2_short(Bulk = Bulk,  
                                                            Background = Background,
                                                            Rep = Dataset,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Background +  (1 | Rep)"),
                                                                  label = c("cept","Bulk", "Background", "Interaction"))  -> CuSO4_chr1_glmer_1byRep_PERM

saveRDS(CuSO4_chr1_glmer_1byRep_PERM, "Output/CuSO4_chr1_glmer_1byRep_PERMG.rds")

Perm1contrasts %>%
  filter(Pool %in% c("HVYTYDRX2","HTG3TDMXY")) %>% distinct() %>% 
  mutate(POS = as.numeric(POS)) %>%
  mutate_if(is.character, as.factor) %>%
  group_by(CHROM, POS) %>% reframe(zscore = glmer_cb2_short(Bulk = Bulk,  
                                                            Background = Background,
                                                            Rep = Dataset,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Background + (1 | Rep)"),
                                                                  label = c("cept","Bulk", "Background", "Interaction"))  -> CuSO4_chr8_glmer_1byRep_PERM

saveRDS(CuSO4_chr8_glmer_1byRep_PERM, "Output/CuSO4_chr8_glmer_1byRep_PERMG.rds")

```

Find 5% FDR by quantile for each coefficient
```{r}
CuSO4_chr8_glmer_1byRep_PERM %>% na.omit() %>% group_by(label) %>% 
  summarize(q5 = quantile(abs(zscore), 0.95)) %>% mutate(chr = "VIII") -> q5_chr8

CuSO4_chr1_glmer_1byRep_PERM %>% na.omit() %>% group_by(label) %>% 
  summarize(q5 = quantile(abs(zscore), 0.95)) %>% mutate(chr = "I") %>% 
  rbind(q5_chr8) -> q5_CuSO4

saveRDS(q5_CuSO4, file = "Output/q5_CuSO4.rds")

head(q5_CuSO4)

rm(q5_chr8)
```

## Combine datasets for analysis

```{r}
CuSO4_chr1_glmer_1byRep_PERM %>% mutate(chr = "I") -> c1
CuSO4_chr8_glmer_1byRep_PERM %>% mutate(chr = "VIII") -> c8

rbind(c1, c8) %>%
  merge(q5_CuSO4) -> CuSO4_glmer_output

saveRDS(CuSO4_glmer_output, file = "CuSO4_glmer_output.rds")

write.csv(CuSO4_glmer_output, file = "CuSO4_glmer_output.csv")

head(CuSO4_glmer_output)

```

